anta/catalog.py

```diff
@@ -0,0 +1,274 @@
+# Copyright (c) 2023 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+"""
+Catalog related functions
+"""
+from __future__ import annotations
+
+import importlib
+import logging
+from inspect import isclass
+from types import ModuleType
+from typing import Any, Dict, List, Type
+
+from pydantic import BaseModel, ConfigDict, RootModel, ValidationInfo, field_validator, model_serializer, model_validator
+from pydantic.types import ImportString
+from yaml import YAMLError, safe_load
+
+from anta.models import AntaTest
+
+logger = logging.getLogger(__name__)
+
+
+class AntaTestDefinition(BaseModel):
+    """
+    Define a test with its associated inputs.
+
+    test: An AntaTest concrete subclass
+    inputs: The associated AntaTest.Input subclass instance
+    """
+
+    model_config = ConfigDict(frozen=True)
+
+    test: Type[AntaTest]
+    inputs: AntaTest.Input
+
+    def __init__(self, **data: Any) -> None:
+        """
+        Inject test in the context to allow to instantiate Input in the BeforeValidator
+        https://docs.pydantic.dev/2.0/usage/validators/#using-validation-context-with-basemodel-initialization
+        """
+        self.__pydantic_validator__.validate_python(
+            data,
+            self_instance=self,
+            context={"test": data["test"]},
+        )
+        super(BaseModel, self).__init__()
+
+    @model_serializer
+    def ser_model(self) -> Dict[str, AntaTest.Input]:
+        """
+        Serialize an AntaTestDefinition as it is defined in a test catalog YAML file.
+        """
+        return {self.test.__name__: self.inputs}
```

Comment:
`self.inputs` should be serialized too no?

Reply:
Yes, actually this method is not used or tested yet.
The idea is to be able to dump a YAML file from an `AntaCatalog` object.
The output we get from `anta check catalog` comes from the __repr__() and is not a serialisation.

So we either delete it or leave it as is with a TODO.